code/modules/reagents/chemistry/holder.dm

```diff
@@ -898,24 +898,85 @@
 	reagents = new /datum/reagents(max_vol, flags)
 	reagents.my_atom = src
 
-/proc/get_random_reagent_id()	// Returns a random reagent ID minus blacklisted reagents and most foods and drinks
-	var/static/list/random_reagents = list()
-	if(!random_reagents.len)
-		for(var/thing  in subtypesof(/datum/reagent))
-			var/datum/reagent/R = thing
-			if(initial(R.can_synth) && initial(R.random_unrestricted))
-				random_reagents += R
-	var/picked_reagent = pick(random_reagents)
-	return picked_reagent
+/proc/get_random_reagent_id(var/flag_check)	// Returns a random reagent ID based on flag_check of each chemicals
+	var/picked_reagent = null
+
+	switch(flag_check) // We check flag first. check 'code\__DEFINES\reagents.dm' for the flags.
+		if(CHEMICAL_RNG_GENERAL) // this is general rng flag stuff
+			var/static/list/random_reagents_a = list() // static list (variable names should be different for each flag)
+			if(!random_reagents_a.len) // if static list has nothing
+				for(var/each in subtypesof(/datum/reagent)) // get all reagents
```

Comment:
`for(var/datum/reagent/R in subtypesof(/datum/reagent))` doesn't work. Byond magic.